tests/pytest/test_logic.py

```diff
@@ -0,0 +1,286 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+from cocotb.types import Logic, Bit
+import pytest
+
+
+def test_logic_conversions():
+    Logic(0)
+    Logic(1)
+    Logic(True)
+    Logic(False)
+    Logic('0')
+    Logic('l')
+    Logic('L')
+    Logic('1')
+    Logic('h')
+    Logic('H')
+    Logic('x')
+    Logic('X')
+    Logic('w')
+    Logic('W')
+    Logic('u')
+    Logic('U')
+    Logic('-')
+    Logic('z')
+    Logic('Z')
+    Logic(Logic('0'))
+    Logic(Logic('1'))
+    Logic(Logic('X'))
+    Logic(Logic('Z'))
+
+    for value in ('j', 2, object()):
+        with pytest.raises(ValueError):
+            Logic(value)
+
+
+def test_bit_conversions():
+    Bit(0)
+    Bit(1)
+    Bit(False)
+    Bit(True)
+    Bit('0')
+    Bit('1')
+    Bit(Bit('0'))
+    Bit(Bit('1'))
+
+    for value in ('X', 2, object()):
+        with pytest.raises(ValueError):
+            Bit(value)
+
+
+def test_bit_logic_conversions():
+    Logic(Bit(0))
+    Logic(Bit(1))
+    Bit(Logic(0))
+    Bit(Logic(1))
+    with pytest.raises(ValueError):
+        Bit(Logic('X'))
+    with pytest.raises(ValueError):
+        Bit(Logic('Z'))
+
+
+def test_logic_equality():
+    assert Logic(0) == Logic('0')
+    assert Logic(0) != Logic('X')
+    assert Logic(0) != object()
+
+
+def test_bit_equality():
+    assert Bit(0) == Bit(False)
+    assert Bit(1) != Bit('0')
+    assert Bit(1) != object()
+
+
+def test_logic_bit_equality():
+    assert Logic(0) != Bit(0)
+    assert Logic(1) != Bit(1)
+
+
+def test_logic_self_identity():
+    l = Logic('0')
+    assert Logic('l') == l
+    assert Logic('L') == l
+    assert Logic(0) == l
+    assert Logic(False) == l
+    assert Logic(Logic('0')) == l
```

Comment:
If you're testing `identity` you should use `is` not `==`

Reply:
I think its just poorly named. I don't want `is` in the test, its an implementation detail. I just want to show that certain literals all result in the same value.

Reply:
> its an implementation detail.

Are you sure we don't want to declare it part of the API? For instance, python's `Enum` declares it part of the API, https://docs.python.org/3/library/enum.html#comparisons

> Enumeration members are compared by identity

And IMO `Logic` is just a glorified enum with some bitwise-ish operators.

Reply:
> Are you sure we don't want to declare it part of the API?

Not sure why we would want to. I know they do, but I don't see the value. `is` would work like `==` and I think that's it. I don't want people to do that either.

> And IMO Logic is just a glorified enum with some bitwise-ish operators.

Exactly. That's how it's defined in both Verilog and VHDL. There isn't much to it. Most of the magic happens in the array of logic specialization, which is yet to come.

Reply:
> I know they do, but I don't see the value

IMO consistency with design decisions in Python is an argument in itself, especially in the absence of counter-arguments.
As a bonus, `is` is faster than `==`.

> Exactly. That's how it's defined in both Verilog and VHDL.

Sorry, to be clear, my point is since we agree that `Logic` is a glorified enum, it makes sense to match the bits of the Python API that don't get in the way.

Reply:
> IMO consistency with design decisions in Python is an argument in itself, especially in the absence of counter-arguments.

Many Python leaders have argued against using `is` for comparsion, and Python even throws a `SyntaxWarning` now if it can detect it. In general having a smaller interface is better than a larger interface. There is also nothing like `is` in Verilog or VHDL.

> As a bonus, `is` is faster than `==`.

Not in this case since they are equivalent. I tested.

Reply:
This is no longer an issue, `Bit` is now a proper subtype and the singleton cache is gone, using `is` here won't work.

Reply:
> Many Python leaders have argued against using is for comparsion

Only for built-in types not designed to use `is` for comparison like str and int.

What was the conclusion here - are values of `bit` singletons like `True` and `False` are?